compiler/rustc_middle/src/ty/sty.rs

```diff
@@ -388,9 +390,17 @@ impl<'tcx> ClosureSubsts<'tcx> {
         self.split().parent_substs
     }
 
+    /// Returns an iterator over the list of types of captured paths by the closure.
+    /// In case there was a type error in figuring out the types of the captured path, an
+    /// empty iterator is returned.
     #[inline]
     pub fn upvar_tys(self) -> impl Iterator<Item = Ty<'tcx>> + 'tcx {
-        self.tupled_upvars_ty().tuple_fields()
+        match self.tupled_upvars_ty().kind() {
+            TyKind::Error(_) => Either::Left(std::iter::empty()),
+            TyKind::Tuple(..) => Either::Right(self.tupled_upvars_ty().tuple_fields()),
```

Comment:
...and `Some(self.tupled_upvars_ty().tuple_fields())` here...

Reply:
Fixed